[ConstraintSystem] Avoid applying invalid solution with fixes

xedin opened the pull request:
If solution score indicates that there should be a fix but
none where recorded fail solution application and produce
a fallback diagnostic to pin-point a problem.

Couple of points:

* Start tracking number of holes involved in each solution, this allows us to:
  a. Score solutions based on number of holes as a fallback for "fix" weight;
  b. Identify cases where there are one or more holes but not "fixes".

* Detect misuse of `_` (discard assignment expression) while performing
  constraint generation which helps to avoid solutions with holes and no fixes
  which used to be diagnosed by syntactic diagnostics before (in `MiscDiagnostics`).

Resolves: rdar://problem/60663007

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Avoid applying invalid solution with fixes

If solution score indicates that there should be a fix but
none where recorded fail solution application and produce
a fallback diagnostic to pin-point a problem.

Resolves: rdar://problem/60663007

xedin committed:
[ConstraintSystem] Score solutions based on number of holes

Introduce `SK_Hole` which is used to count a number of "holes" in
a given solution. It is used to distinguish solutions with fewer holes.

Also it makes it possible to check whether a solution has holes but
no fixes, which is an issue and such solution shouldn't be applied
to AST.

xedin committed:
[ConstraintSystem] Diagnose incorrect use of `_` during constraint generation

`_` or discard assignment expression should only be used on the left-hand
side of the assignment expression. Incorrect uses are easy to detect during
constraint generation which also allows us to avoid complications related
to other diagnostics when `_` is used incorrectly.

xedin committed:
[ConstraintSystem] Produce fallback diagnostic on mismatch between fixes/holes

If score indicates that a solution should have at least one fix
but it doesn't let's attempt to produce a fallback diagnostic inline.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test macOS platform

xedin commented:
@swift-ci please smoke test macOS platform

xedin commented:
@swift-ci please smoke test macOS platform

hborla reviewed (commented):
Looks great! I just have one small comment about using `SK_Hole` in other places

xedin commented on lib/Sema/ConstraintSystem.h:
```diff
@@ -634,6 +634,8 @@ enum ScoreKind {
   /// A fix needs to be applied to the source.
   SK_Fix,
+  /// A hole in the constraint system.
+  SK_Hole,
```
Sure, I'll take a look!

xedin replied:
Done!

xedin committed:
[CSGen] No longer allow `_` allow to be a hole directly

Problems related to incorrect use of `_` where previously diagnosed
by syntactic diagnostics which meant that it could only happen on
type-checked AST. This is no longer a case so we don't have to allow
incorrect uses of `_` to form solution without any fixes.

xedin committed:
[ConstraintSystem] Don't increase `SK_Fix` score when looking through holes

`SK_Fix` was used to indicate that solver has encountered a hole
along the current path but since there is `SK_Hole` now, increasing
`SK_Fix` no longer makes sense.

xedin commented:
@swift-ci please smoke test
